File: insertion-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode *dummy = new ListNode(-500000), *node;
        dummy->next = head;
        node = dummy;
        while(head)
        {
            if(node->val > head->val)
            {
                node->next = head->next;
                ListNode *ptr = dummy;
                while(ptr->next->val <= head->val)
                    ptr = ptr->next;
                head->next = ptr->next;
                ptr->next = head;
                head = node->next;
            }
            else
            {
                node = node->next;
                head = head->next;
            }
        }
        
        return dummy->next;
        
    }
};

File: leetcode/insertion-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummyNode = ListNode(0)
        dummy = dummyNode

        curr = head
        while curr:
            temp = curr.next
            dummy= dummyNode
            while dummy.next and dummy.next.val < curr.val:
                dummy = dummy.next
            if dummy.next == None:
                dummy.next = curr
                curr.next = None
            else:
                nxt = dummy.next
                dummy.next = curr
                curr.next=nxt
            curr = temp
        return dummyNode.next


        